presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/client/AwsRequestSigner.java

```diff
@@ -0,0 +1,115 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.elasticsearch.client;
+
+import com.amazonaws.DefaultRequest;
+import com.amazonaws.auth.AWS4Signer;
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.http.HttpMethodName;
+import org.apache.http.Header;
+import org.apache.http.HttpEntityEnclosingRequest;
+import org.apache.http.HttpHost;
+import org.apache.http.HttpRequest;
+import org.apache.http.HttpRequestInterceptor;
+import org.apache.http.NameValuePair;
+import org.apache.http.client.utils.URIBuilder;
+import org.apache.http.entity.BasicHttpEntity;
+import org.apache.http.message.BasicHeader;
+import org.apache.http.protocol.HttpContext;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.lang.String.CASE_INSENSITIVE_ORDER;
+import static org.apache.http.protocol.HttpCoreContext.HTTP_TARGET_HOST;
+
+class AwsRequestSigner
+        implements HttpRequestInterceptor
+{
+    private static final String SERVICE_NAME = "es";
+    private final AWSCredentialsProvider credentialsProvider;
+    private final AWS4Signer signer;
+
+    public AwsRequestSigner(String region, AWSCredentialsProvider credentialsProvider)
+    {
+        this.credentialsProvider = credentialsProvider;
+        this.signer = new AWS4Signer();
+
+        signer.setServiceName(SERVICE_NAME);
+        signer.setRegionName(region);
+    }
+
+    @Override
+    public void process(final HttpRequest request, final HttpContext context)
+            throws IOException
+    {
+        String method = request.getRequestLine().getMethod();
+
+        URI uri = URI.create(request.getRequestLine().getUri());
+        URIBuilder uriBuilder = new URIBuilder(uri);
+
+        Map<String, List<String>> parameters = new TreeMap<>(CASE_INSENSITIVE_ORDER);
+        for (NameValuePair parameter : uriBuilder.getQueryParams()) {
+            parameters.computeIfAbsent(parameter.getName(), key -> new ArrayList<>())
+                    .add(parameter.getValue());
+        }
+
+        Map<String, String> headers = Arrays.stream(request.getAllHeaders())
+                .collect(toImmutableMap(Header::getName, Header::getValue));
+
+        InputStream content = null;
+        if (request instanceof HttpEntityEnclosingRequest) {
+            HttpEntityEnclosingRequest enclosingRequest = (HttpEntityEnclosingRequest) request;
+            if (enclosingRequest.getEntity() != null) {
+                content = enclosingRequest.getEntity().getContent();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *content*. Currently, there are execution paths that do not contain closure statements, for example, when HttpHost.toURI throws an exception. Ensure *content* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).